p2p/exchange_metrics.go

```diff
@@ -0,0 +1,165 @@
+	trackedPeersNumReg  metric.Registration
+
+	disconnectedPeersNum     atomic.Int64
+	disconnectedPeersNumInst metric.Int64ObservableGauge
+	disconnectedPeersNumReg  metric.Registration
+}
+
+func newExchangeMetrics() (m *exchangeMetrics, err error) {
+	m = new(exchangeMetrics)
+	m.headRequestTimeInst, err = meter.Int64Histogram(
+		"hdr_p2p_exch_clnt_head_time_hist",
+		metric.WithDescription("exchange client head request time in milliseconds"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.responseSizeInst, err = meter.Int64Histogram(
+		"hdr_p2p_exch_clnt_resp_size_hist",
+		metric.WithDescription("exchange client header response size in bytes"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.responseTimeInst, err = meter.Int64Histogram(
+		"hdr_p2p_exch_clnt_resp_time_hist",
+		metric.WithDescription("exchange client response time in milliseconds"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.trackedPeersNumInst, err = meter.Int64ObservableGauge(
+		"hdr_p2p_exch_clnt_trck_peer_num_gauge",
+		metric.WithDescription("exchange client tracked peers number"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.trackedPeersNumReg, err = meter.RegisterCallback(m.observeTrackedPeers, m.trackedPeersNumInst)
+	if err != nil {
+		return nil, err
+	}
+	m.disconnectedPeersNumInst, err = meter.Int64ObservableGauge(
+		"hdr_p2p_exch_clnt_disc_peer_num_gauge",
+		metric.WithDescription("exchange client tracked disconnected peers number"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.disconnectedPeersNumReg, err = meter.RegisterCallback(m.observeDisconnectedPeers, m.disconnectedPeersNumInst)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *exchangeMetrics) head(ctx context.Context, duration time.Duration, headersReceived int, tp, status string) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.headRequestTimeInst.Record(ctx,
+			duration.Milliseconds(),
+			metric.WithAttributes(
+				attribute.Int(headerReceivedKey, headersReceived),
+				attribute.String(headTypeKey, tp),
+				attribute.String(headStatusKey, status),
+			),
+		)
+	})
+}
+
+func (m *exchangeMetrics) response(ctx context.Context, size uint64, duration time.Duration, err error) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.responseSizeInst.Record(ctx,
+			int64(size),
+			metric.WithAttributes(attribute.Bool(failedKey, err != nil)),
+		)
+		m.responseTimeInst.Record(ctx,
+			duration.Milliseconds(),
+			metric.WithAttributes(attribute.Bool(failedKey, err != nil)),
+		)
+	})
+}
+
+func (m *exchangeMetrics) peersTracked(num int) {
+	m.observe(context.Background(), func(context.Context) {
+		m.trackerPeersNum.Add(int64(num))
+	})
+}
+
+func (m *exchangeMetrics) peersDisconnected(num int) {
+	m.observe(context.Background(), func(context.Context) {
+		m.disconnectedPeersNum.Add(int64(num))
+	})
+}
+
+func (m *exchangeMetrics) peerBlocked(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.blockedPeersNum.Add(ctx, 1)
```

Comment:
since this shouldn't be high cardinality, can we add attribute peer ID here?

Reply:
Hmm, interesting. I think this makes sense

Reply:
However, there is no way to query those, as per @walldiss . Attributes have to be predefined and you need to know them in advance.

Reply:
we already have an RPC for blocked nodes on the host but we would need to know where they got "blocked from" (so whether the blocking happened in peerTracker or in shrexsub for some reason or in befp

Reply:
Then we are good